tests/test_visitors.py

```diff
@@ -205,3 +237,73 @@ def test_transformer_add_replace(exprs, block2, block3):
         assert line1 in newcode
         assert line2 in newcode
         assert "a[i0] = a[i0] + b[i0] + 5.0F;" not in newcode
+
+
+def test_merge_iterations_flat(exprs, iters):
+    """Test outer loop merging on a simple two-level hierarchy:
+
+    for i                       for i
+        for j              \        for j
+            expr0       === \           expr0
+    for i               === /       for k
+        for k              /            expr1
+            expr1
```

Comment:
Note: it's always legal to merge these two trees as long as ``expr0`` is not a reduction along ``for i`` *AND* ``expr0.lhs`` doesn't appear in ``expr1.rhs``. By reduction "along for i" I mean something like:

```
for i
  for j
    a[j] += c[j]*3
```

The DSE provides the ``TemporariesGraph``, an object that could be used to this purpose. I'd leave a comment in ``StencilKernel`` to remember to introduce this check prior to merging.